broken package opencv

Commit 697251a497583ed446012eea90a4b55f509cf4cc
Target: x86_64-w64-mingw32.static

[Log](https://gist.githubusercontent.com/starius/dce00dc67dbc2e0827c3/raw/opencv_x86_64-w64-mingw32.static)

Currently `opencv` is broken by `libjpeg-turbo` (header `include/jmorecfg.h`). See #826 for details.

There are also linking problems (see the log from my first comment) for x86_64-w64-mingw32.static:

```
'/home/mxe/mxe-db-enable-crypto/usr/bin/x86_64-w64-mingw32.static-g++'
-W -Wall -Werror -ansi -pedantic 
'/home/mxe/mxe-db-enable-crypto/tmp-opencv-x86_64-w64-mingw32.static/o
pencv-2.4.10/samples/c/fback_c.c' -o 
'/home/mxe/mxe-db-enable-crypto/usr/x86_64-w64-mingw32.static/bin/test
-opencv.exe' 
`'/home/mxe/mxe-db-enable-crypto/usr/bin/x86_64-w64-mingw32.static-pkg
-config' opencv --cflags --libs`
```

This is expanded to
```
'/home/mxe/mxe-db-enable-crypto/usr/bin/x86_64-w64-mingw32.static-g++'
-W -Wall -Werror -ansi -pedantic 
'/home/mxe/mxe-db-enable-crypto/tmp-opencv-x86_64-w64-mingw32.static/o
pencv-2.4.10/samples/c/fback_c.c' -o 
'/home/mxe/mxe-db-enable-crypto/usr/x86_64-w64-mingw32.static/bin/test
-opencv.exe' 
-I/home/mxe/mxe-db-enable-crypto/usr/x86_64-w64-mingw32.static/include
/opencv -lopencv_contrib -lopencv_stitching -lopencv_nonfree 
-lopencv_superres -lopencv_ocl -lopencv_ts -lopencv_videostab 
-lopencv_gpu -lopencv_photo -lopencv_objdetect -lopencv_legacy 
-lopencv_video -lopencv_ml -lopencv_calib3d -lopencv_features2d 
-lopencv_highgui -lopencv_imgproc -lopencv_flann -lopencv_core 
-lmsvfw32 -lwinmm -lavicap32 -lavifil32 -lstrmiids -lws2_32 
-lsetupapi -lole32 -lgdi32 -lcomctl32 -lIlmThread -lHalf -lIex 
-lIlmImf -lImath -ljasper -lpng -lopengl32 -lglu32 -lstdc++ -ltiff 
-luuid -loleaut32 -llzma -ljpeg -lz
```

This command fails because of wrong order of libraries.

```
/usr/lib/mxe/usr/lib/gcc/x86_64-w64-mingw32.static/5.1.0/../../../../x
86_64-w64-mingw32.static/lib/../lib/libIlmImf.a(ImfChannelList.o):ImfC
hannelList.cpp:(.text+0xe3): undefined reference to `iex_debugTrap()'
...
/usr/lib/mxe/usr/lib/gcc/x86_64-w64-mingw32.static/5.1.0/../../../../x
86_64-w64-mingw32.static/lib/../lib/libIlmImf.a(ImfHeader.o):ImfHeader
.cpp:(.text+0x397d): undefined reference to 
`IlmThread_2_2::Mutex::~Mutex()'
...
```

This error message (and grep in usr/x86_64-w64-mingw32.static/lib) tells us that:
  * `IlmImf` depends on `Iex`
  * `IlmImf` depends on `IlmThread`

To fix it, move `-lIex -lIlmThread` after `IlmImf`:

```
'/home/mxe/mxe-db-enable-crypto/usr/bin/x86_64-w64-mingw32.static-g++'
-W -Wall -Werror -ansi -pedantic 
'/home/mxe/mxe-db-enable-crypto/tmp-opencv-x86_64-w64-mingw32.static/o
pencv-2.4.10/samples/c/fback_c.c' -o 
'/home/mxe/mxe-db-enable-crypto/usr/x86_64-w64-mingw32.static/bin/test
-opencv.exe' 
-I/home/mxe/mxe-db-enable-crypto/usr/x86_64-w64-mingw32.static/include
/opencv -lopencv_contrib -lopencv_stitching -lopencv_nonfree 
-lopencv_superres -lopencv_ocl -lopencv_ts -lopencv_videostab 
-lopencv_gpu -lopencv_photo -lopencv_objdetect -lopencv_legacy 
-lopencv_video -lopencv_ml -lopencv_calib3d -lopencv_features2d 
-lopencv_highgui -lopencv_imgproc -lopencv_flann -lopencv_core 
-lmsvfw32 -lwinmm -lavicap32 -lavifil32 -lstrmiids -lws2_32 
-lsetupapi -lole32 -lgdi32 -lcomctl32 -lHalf -lIlmImf -lIex 
-lIlmThread -lImath -ljasper -lpng -lopengl32 -lglu32 -lstdc++ 
-ltiff -luuid -loleaut32 -llzma -ljpeg -lz
```
Now it works.

To fix this issue, I investigate file `cmake/OpenCVGenPkgconfig.cmake`. It defines complicated logic of generating `opencv.pc`. Maybe the issue is cause by error in this code:

```cmake
#build the list of opencv libs and dependencies for all modules
set(OpenCV_LIB_COMPONENTS "")
set(OpenCV_EXTRA_COMPONENTS "")
foreach(m ${OPENCV_MODULES_PUBLIC})
  list(INSERT OpenCV_LIB_COMPONENTS 0 ${${m}_MODULE_DEPS_${ocv_optkind}} ${m})
  if(${m}_EXTRA_DEPS_${ocv_optkind})
    list(INSERT OpenCV_EXTRA_COMPONENTS 0 ${${m}_EXTRA_DEPS_${ocv_optkind}})
  endif()
endforeach()

ocv_list_unique(OpenCV_LIB_COMPONENTS)
ocv_list_unique(OpenCV_EXTRA_COMPONENTS)
ocv_list_reverse(OpenCV_LIB_COMPONENTS)
ocv_list_reverse(OpenCV_EXTRA_COMPONENTS)
```

Variable `OpenCV_LIB_COMPONENTS` is inserted into `opencv.pc` (see file `cmake/templates/opencv-XXX.pc.in`):
```
# Package Information for pkg-config

prefix=@prefix@
exec_prefix=@exec_prefix@
libdir=@libdir@
includedir_old=@includedir@/opencv
includedir_new=@includedir@

Name: OpenCV
Description: Open Source Computer Vision Library
Version: @OPENCV_VERSION@
Libs: @OpenCV_LIB_COMPONENTS@
Cflags: -I${includedir_old} -I${includedir_new}
```

--- a/src/opencv-4-reverse-deps-sublist.patch
+++ b/src/opencv-4-reverse-deps-sublist.patch
@@ -1,0 +1,38 @@
+This file is part of MXE.
+See index.html for further information.
+
+From 20cf852345e364f7e18c8721561d7c4d1ebd1d6e Mon Sep 17 00:00:00 2001
+From: Boris Nagaev <bnagaev@gmail.com>
+Date: Sun, 6 Sep 2015 00:53:13 +0100
+Subject: [PATCH] reverse deps list before inserting to another list
+
+Order of items in ${m}_EXTRA_DEPS_${ocv_optkind} corresponds
+to the order in which a linker expects items to appear:
+a depender before a dependee.
+
+But the order of items of OpenCV_EXTRA_COMPONENTS is
+reversed (afterwards this list is unique'ed and reversed).
+That is why sub-lists must be inserted in reversed form.
+---
+ cmake/OpenCVGenPkgconfig.cmake | 4 ++++
+ 1 file changed, 4 insertions(+)
+
+diff --git a/cmake/OpenCVGenPkgconfig.cmake b/cmake/OpenCVGenPkgconfig.cmake
+index 183c56d..180210d 100644
+--- a/cmake/OpenCVGenPkgconfig.cmake
++++ b/cmake/OpenCVGenPkgconfig.cmake
+@@ -19,9 +19,11 @@ endif()
+ set(OpenCV_LIB_COMPONENTS "")
+ set(OpenCV_EXTRA_COMPONENTS "")
+ foreach(m ${OPENCV_MODULES_PUBLIC})
+   list(INSERT OpenCV_LIB_COMPONENTS 0 ${${m}_MODULE_DEPS_${ocv_optkind}} ${m})
+   if(${m}_EXTRA_DEPS_${ocv_optkind})
++    ocv_list_reverse(${m}_EXTRA_DEPS_${ocv_optkind})
+     list(INSERT OpenCV_EXTRA_COMPONENTS 0 ${${m}_EXTRA_DEPS_${ocv_optkind}})
++    ocv_list_reverse(${m}_EXTRA_DEPS_${ocv_optkind})
+   endif()
+ endforeach()
+ 
+-- 
+2.1.4
+

